os/kernel/task/task_exithook.c

```diff
@@ -566,6 +571,14 @@ void task_exithook(FAR struct tcb_s *tcb, int status, bool nonblocking)
 		return;
 	}
 
+#ifdef CONFIG_TASK_MANAGER
+	/* If tm_cancel_flag is TM_IS_CANCELLED, then callback is already called. */
+	if (tm_cancel_flag != TM_IS_CANCELLED) {
```

Comment:
1. Let's check this flag inside ```task_manager_run_exit_cb```.
2. Let's move this logic under cancallation, above atexit.

Reply:
I think that checking flag in here is better, because if move to task_manager_run_exit_cb inside, we always call that function.

Reply:
I'm wondering multiple threads may enter this if statement. If so, simple flag check doesn't protect this critical section, but a semaphore or mutex may be required.